llvm/lib/Target/DirectX/DXILOpLowering.cpp

```diff
@@ -265,16 +265,50 @@ class OpLowerer {
 
   /// Replace uses of \c Intrin with the values in the `dx.ResRet` of \c Op.
   /// Since we expect to be post-scalarization, make an effort to avoid vectors.
-  Error replaceResRetUses(CallInst *Intrin, CallInst *Op) {
+  Error replaceResRetUses(CallInst *Intrin, CallInst *Op, bool HasCheckBit) {
     IRBuilder<> &IRB = OpBuilder.getIRB();
 
+    Instruction *OldResult = Intrin;
     Type *OldTy = Intrin->getType();
 
+    if (HasCheckBit) {
+      auto *ST = cast<StructType>(OldTy);
```

Comment:
Does the success of cast need to be ensured by checking `ST != nullptr`, here and similarly on line 299?

Reply:
`cast` asserts if it can't do the cast (unlike `dyn_cast`, which returns null). It's meant to be used only when it's known that it can't fail.